### DISCUSSION TITLE
Multiclass interpration questions / maybe issues with SHAPley

### DISCUSSION THREAD

**Discussion by Severac** [shap/shap#1588] [Q&A]
Hi,

I am trying to interpret individual predictions of a XGBoost model with the SHAPLEY python library.
This model is multi-class (but single label).

Unfortunately I could not reproduce most of the problems with a toy dataset and code, and I cannot paste my original code since the data is confidential and too big.
But I was wondering if you could have insights on this, though I'm sorry not to have reproductible code.

Still, here's the toy code :
https://www.kaggle.com/franoisboyer/machine-learning-with-small-dataset-shapley-test :

I have 3 interrogations / problems :

1/ When I call the equivalent of this in my original code :

    shap_values = explainer.shap_values(X_train[idx, :].reshape(1, 35))

I have the following exception :

> Exception: The additivity check failed in TreeExplainer! Please ensure the
> data matrix you passed to the explainer is the same shape that the
> model was trained on. If your data shape is correct then please report
> this on GitHub. Consider retrying with the
> feature_perturbation='interventional' option. This check failed
> because for one of the samples the sum of the SHAP values was
> -0.188287, while the model output was -0.110077. If this difference is acceptable you can set check_additivity=False to disable this check.

=> Can this be normal or is it always a problem?  
=> With a reduced subset of training data passed to fit the function of the model, I sometimes did not have this exception message  
=> I don't have this problem with the toy code  

2/ When I call this with check_additivity = False :

    shap_values = explainer.shap_values(X_train[idx, :].reshape(1, 35), check_additivity=False)

The results seem ok for some predictions,
But for some other predictions, the highest of the 3 shap values  (one shap value per class, and I have 3 classes) does not match with the predicted class :

For example below: predicted class is "Refus" but the highest shap value matchs "Accord total" class

Image attached
![Shap force plot](https://i.imgur.com/cjzjo1p.png)
(variables are hidden in this screenshot)

=> Is it normal/possible behavior for multi-class models ? Maybe shap scale is not always consistent between different classes, which could lead to this for some predictions (but not all)?  
=> I don't have this problem with the toy code  
=> It's surprising to see that the -7.23 value appears more to the right, after -0.06 value: maybe that's because -7.23 corresponds to the predicted class, but this is not always the case: the value corresponding to the predicted class does not always appear on the far right  (although it does for most predictions...)  

3/
Is it normal that the 2 following values are not equal (I have the same problem in toy code) :

    print('Prediction:')
    print(model.predict_proba(X_train[idx, :].reshape(1, 35))[0].round(3))
    print('Expected value + shap values:')
    print((explainer.expected_value + shap_values[0].sum()).round(3))

I've read somewhere that they should be equal.
I also tried to pass link='logit' to shap.force_plot method: that changes the shape values, but the 2 values are still not equal and behavior is the same.

Apart from these problems, positive and negative top features displayed by force_plot seem consistent and helpful for what I've checked so far.
But I'm unsure about things above.

Thanks a lot !
